File: src/grove/state.py
"""Workspace state management (~/.grove/state.json)."""

from __future__ import annotations

import contextlib
import json
import os
import tempfile
from pathlib import Path

from grove.config import GROVE_DIR
from grove.models import Workspace

STATE_PATH = GROVE_DIR / "state.json"


def _load_raw() -> list[dict]:
    """Load raw state data."""
    if not STATE_PATH.exists():
        return []
    try:
        return json.loads(STATE_PATH.read_text())
    except json.JSONDecodeError as e:
        raise SystemExit(
            f"State file is corrupt ({STATE_PATH}): {e}\n"
            "Delete the file to reset, or run: gw doctor --fix"
        ) from e


def _atomic_write(path: Path, content: str) -> None:
    """Write *content* to *path* atomically via temp file + rename."""
    fd, tmp = tempfile.mkstemp(dir=path.parent, suffix=".tmp")
    try:
        with os.fdopen(fd, "w") as f:
            f.write(content)
        os.replace(tmp, path)
    except BaseException:
        with contextlib.suppress(OSError):
            os.unlink(tmp)
        raise


def _save_raw(data: list[dict]) -> None:
    """Save raw state data atomically (write to temp, then replace)."""
    GROVE_DIR.mkdir(parents=True, exist_ok=True)
    _atomic_write(STATE_PATH, json.dumps(data, indent=2) + "\n")


def load_workspaces() -> list[Workspace]:
    """Load all workspaces from state."""
    return [Workspace.from_dict(d) for d in _load_raw()]


def save_workspaces(workspaces: list[Workspace]) -> None:
    """Save all workspaces to state."""
    _save_raw([w.to_dict() for w in workspaces])


def get_workspace(name: str) -> Workspace | None:
    """Get a workspace by name."""
    for ws in load_workspaces():
        if ws.name == name:
            return ws
    return None


def add_workspace(workspace: Workspace) -> None:
    """Add a workspace to state."""
    workspaces = load_workspaces()
    workspaces.append(workspace)
    save_workspaces(workspaces)


def remove_workspace(name: str) -> None:
    """Remove a workspace from state."""
    workspaces = [w for w in load_workspaces() if w.name != name]
    save_workspaces(workspaces)


def update_workspace(workspace: Workspace, *, match_name: str | None = None) -> None:
    """Replace an existing workspace in state.

    Matches by *match_name* if given, otherwise by ``workspace.name``.
    This allows atomic renames (match old name, write new name) in one operation.
    """
    key = match_name or workspace.name
    workspaces = load_workspaces()
    workspaces = [workspace if w.name == key else w for w in workspaces]
    save_workspaces(workspaces)


def find_workspace_by_path(path: Path) -> Workspace | None:
    """Find a workspace that contains the given path."""
    resolved = path.resolve()
    for ws in load_workspaces():
        ws_resolved = ws.path.resolve()
        if resolved == ws_resolved or ws_resolved in resolved.parents:
            return ws
    return None

File: grove-go/internal/state/state.go
// Package state manages the workspace state file (~/.grove/state.json).
package state

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/nicksenap/grove/internal/config"
	"github.com/nicksenap/grove/internal/models"
)

// StatePath returns the path to ~/.grove/state.json.
func StatePath() string {
	return filepath.Join(config.GroveDir(), "state.json")
}

// LoadWorkspaces reads all workspaces from state.json.
func LoadWorkspaces() ([]models.Workspace, error) {
	data, err := os.ReadFile(StatePath())
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}

	var workspaces []models.Workspace
	if err := json.Unmarshal(data, &workspaces); err != nil {
		return nil, fmt.Errorf("parsing state: %w", err)
	}
	return workspaces, nil
}

// SaveWorkspaces atomically persists the workspace list.
func SaveWorkspaces(workspaces []models.Workspace) error {
	data, err := json.MarshalIndent(workspaces, "", "  ")
	if err != nil {
		return err
	}
	return atomicWrite(StatePath(), string(data)+"\n")
}

// GetWorkspace returns a workspace by name, or nil if not found.
func GetWorkspace(name string) (*models.Workspace, error) {
	workspaces, err := LoadWorkspaces()
	if err != nil {
		return nil, err
	}
	lower := strings.ToLower(name)
	for i := range workspaces {
		if strings.ToLower(workspaces[i].Name) == lower {
			return &workspaces[i], nil
		}
	}
	return nil, nil
}

// AddWorkspace appends a workspace to state.
func AddWorkspace(ws models.Workspace) error {
	workspaces, err := LoadWorkspaces()
	if err != nil {
		return err
	}
	workspaces = append(workspaces, ws)
	return SaveWorkspaces(workspaces)
}

// RemoveWorkspace removes a workspace by name from state.
func RemoveWorkspace(name string) error {
	workspaces, err := LoadWorkspaces()
	if err != nil {
		return err
	}
	lower := strings.ToLower(name)
	filtered := workspaces[:0]
	for _, ws := range workspaces {
		if strings.ToLower(ws.Name) != lower {
			filtered = append(filtered, ws)
		}
	}
	return SaveWorkspaces(filtered)
}

// UpdateWorkspace replaces a workspace in state. If matchName is non-empty,
// it matches by that name instead (for renames).
func UpdateWorkspace(ws models.Workspace, matchName string) error {
	workspaces, err := LoadWorkspaces()
	if err != nil {
		return err
	}

	target := ws.Name
	if matchName != "" {
		target = matchName
	}
	lower := strings.ToLower(target)

	found := false
	for i := range workspaces {
		if strings.ToLower(workspaces[i].Name) == lower {
			workspaces[i] = ws
			found = true
			break
		}
	}
	if !found {
		return fmt.Errorf("workspace %q not found", target)
	}
	return SaveWorkspaces(workspaces)
}

// FindWorkspaceByPath returns the workspace whose path contains the given directory.
func FindWorkspaceByPath(dir string) (*models.Workspace, error) {
	workspaces, err := LoadWorkspaces()
	if err != nil {
		return nil, err
	}
	abs, err := filepath.Abs(dir)
	if err != nil {
		return nil, err
	}

	for i := range workspaces {
		wsAbs, _ := filepath.Abs(workspaces[i].Path)
		if strings.HasPrefix(abs, wsAbs+string(filepath.Separator)) || abs == wsAbs {
			return &workspaces[i], nil
		}
	}
	return nil, nil
}

// atomicWrite writes content to a temp file then renames to target.
func atomicWrite(path, content string) error {
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	tmp, err := os.CreateTemp(dir, ".grove-state-*")
	if err != nil {
		return err
	}
	tmpName := tmp.Name()

	if _, err := tmp.WriteString(content); err != nil {
		tmp.Close()
		os.Remove(tmpName)
		return err
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpName)
		return err
	}
	return os.Rename(tmpName, path)
}
